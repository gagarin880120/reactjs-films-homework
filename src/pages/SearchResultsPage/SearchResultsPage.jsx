import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import Header from '../../components/Header/Header';
import MovieListContainer from '../../components/MovieList';
import FooterContainer from '../../components/Footer';
import Spinner from '../../components/Spinner/Spinner';
import MovieInfoContainer from '../../components/MovieInfo';
import styles from './SearchResultsPage.module.scss';

export default function SearchResultsPage({
  results, areMoviesLoaded, isMovieLoaded, currentAPIRequest, onCurrentAPIRequestUpdate,
}) {
  useEffect(() => {
    onCurrentAPIRequestUpdate(currentAPIRequest);
  }, [currentAPIRequest]);
  return (
    <>
      {
        isMovieLoaded ? (
          <>
            <Header />
            <MovieInfoContainer />
          </>
        ) : null
      }
      <MovieListContainer />
      <div className={styles.loading}>
        {
          !areMoviesLoaded ? <Spinner /> : null
        }
      </div>
      {
        results.length ? <FooterContainer /> : null
      }
    </>
  );
}

SearchResultsPage.propTypes = {
  results: PropTypes.instanceOf(Array),
  areMoviesLoaded: PropTypes.bool,
  currentAPIRequest: PropTypes.string,
  onCurrentAPIRequestUpdate: PropTypes.func,
  isMovieLoaded: PropTypes.bool,
};

SearchResultsPage.defaultProps = {
  results: [],
  areMoviesLoaded: false,
  currentAPIRequest: '',
  onCurrentAPIRequestUpdate: null,
  isMovieLoaded: false,
};
