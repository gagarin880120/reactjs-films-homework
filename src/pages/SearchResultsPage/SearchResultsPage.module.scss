.wrapper {
  background: #f5f5f5;
}

.loading {
  height: 60px;
}
