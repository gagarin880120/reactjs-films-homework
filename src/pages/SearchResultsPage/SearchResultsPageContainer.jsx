import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import SearchResultsPage from './SearchResultsPage';
import { resultsSelector, isLoadedSelector } from '../../redux/selectors';
import { getMovies } from '../../redux/actions';
import getURL from '../../utils/utils';

export function SearchResultsPageContainer({ onTrending, results, isLoaded }) {
  useEffect(() => {
    onTrending();
  }, []);

  return (
    <SearchResultsPage
      results={results}
      isLoaded={isLoaded}
    />
  );
}

export const mapStateToProps = (state) => ({
  results: resultsSelector(state),
  isLoaded: isLoadedSelector(state),
})

export const mapDispatchToProps = (dispatch) => ({
  onTrending() {
    dispatch(getMovies(getURL('trending'), 1));
  },
});

SearchResultsPageContainer.propTypes = {
  onTrending: PropTypes.func,
  results: PropTypes.instanceOf(Array),
  isLoaded: PropTypes.bool,
};

SearchResultsPageContainer.defaultProps = {
  onTrending: null,
  results: [],
  isLoaded: false,
};

export default connect(mapStateToProps, mapDispatchToProps)(SearchResultsPageContainer);
