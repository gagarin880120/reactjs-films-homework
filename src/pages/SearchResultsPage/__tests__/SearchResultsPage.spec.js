import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import TestRenderer, { act } from 'react-test-renderer';
import SearchResultsPage from '../SearchResultsPage';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';

jest.mock('react-router-dom', () => ({
  Link: () => null,
  useHistory: jest.fn().mockReturnValue({
    location: { pathname: '/popular' },
  }),
}));

describe('SearchResultsPage component', () => {
  test('should render without crashing', () => {
    const renderer = new ShallowRenderer();
    renderer.render(
      <SearchResultsPage areMoviesLoaded={true} isMovieLoaded={true}/>);
    const result = renderer.getRenderOutput();
    expect(result).toMatchSnapshot();
  });

  test('should render with other elements, depends on props', () => {
    const renderer = new ShallowRenderer();
    renderer.render(
      <SearchResultsPage areMoviesLoaded={false} results={[1, 2, 3]} />);
    const result = renderer.getRenderOutput();
    expect(result).toMatchSnapshot();
  });

  describe('mounted with TestRenderer', () => {
    const middlewares = [thunk];
    const mockStore = configureMockStore(middlewares);
    const store = mockStore({
      results: [{title: 'Matrix', genre_ids:[28], id: 398}],
      genres: [{id: 28, name: 'Action'}]
    });
    const onCurrentAPIRequestUpdate = jest.fn();

    const testRenderer = TestRenderer.create(
      <Provider store={store}>
        <SearchResultsPage
          currentAPIRequest="popular"
          onCurrentAPIRequestUpdate={onCurrentAPIRequestUpdate}
        />
      </Provider>
    );

    test('when currentAPIRequest is updated should call onCurrentAPIRequestUpdate', () => {

      act(() => {
        testRenderer.update(
          <Provider store={store}>
            <SearchResultsPage
              currentAPIRequest="search=matrix"
              onCurrentAPIRequestUpdate={onCurrentAPIRequestUpdate}
            />
          </Provider>
        );
      });

      expect(onCurrentAPIRequestUpdate).toHaveBeenCalled();
    });
  });
});
