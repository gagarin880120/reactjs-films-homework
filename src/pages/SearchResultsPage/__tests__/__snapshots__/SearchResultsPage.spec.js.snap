// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchResultsPage component should render with other elements, depends on props 1`] = `
<React.Fragment>
  <Memo(Connect(MovieListContainer)) />
  <div
    className="loading"
  >
    <Spinner />
  </div>
  <Memo(Connect(FooterContainer)) />
</React.Fragment>
`;

exports[`SearchResultsPage component should render without crashing 1`] = `
<React.Fragment>
  <React.Fragment>
    <Header />
    <Memo(Connect(MovieInfoContainer)) />
  </React.Fragment>
  <Memo(Connect(MovieListContainer)) />
  <div
    className="loading"
  />
</React.Fragment>
`;
